modules/core/04-channel/keeper/upgrade.go

```diff
@@ -293,6 +268,130 @@ func (k Keeper) ChanUpgradeAck(
 	return nil
 }
 
+// ChanUpgradeTimeout times out an outstanding upgrade.
+// This should be used by the initialising chain when the counterparty chain has not responded to an upgrade proposal within the specified timeout period.
+func (k Keeper) ChanUpgradeTimeout(
+	ctx sdk.Context,
+	portID, channelID string,
+	counterpartyChannel types.Channel,
+	prevErrorReceipt types.ErrorReceipt,
+	proofCounterpartyChannel,
+	proofErrorReceipt []byte,
+	proofHeight exported.Height,
+) error {
+	channel, found := k.GetChannel(ctx, portID, channelID)
+	if !found {
+		return errorsmod.Wrapf(types.ErrChannelNotFound, "port ID (%s) channel ID (%s)", portID, channelID)
+	}
+
+	if channel.State != types.INITUPGRADE {
+		return errorsmod.Wrapf(types.ErrInvalidChannelState, "channel state is not INITUPGRADE (got %s)", channel.State)
+	}
+
+	upgrade, found := k.GetUpgrade(ctx, portID, channelID)
+	if !found {
+		return errorsmod.Wrapf(types.ErrUpgradeNotFound, "port ID (%s) channel ID (%s)", portID, channelID)
+	}
+
+	connection, found := k.connectionKeeper.GetConnection(ctx, channel.ConnectionHops[0])
```

Comment:
though we aren't that consistent with this all over the place, we typically also check that the connection is OPEN.

Reply:
i guess it would take some more gas, but maybe worth checking...? wdyt

Reply:
I'd vote for consistency. I remember from a talk with Colin that given that the channel is OPEN, the connection will be too, these were added in the initial channel opening handshake handlers when optimistic packet sends were a possibility (which isn't something we clearly support now afaik).

I guess we could decide at a later point if removing them from certain handlers makes sense.

Reply:
shouldn't cost any extra gas? accessing the store will consume gas but once its retrieved there's not gonna be any gas metering on checking its state

Reply:
ah yea, of course since we already have the connection